### PR TITLE
[#346] Add changelog and bump version

### PR DESCRIPTION
**Pull request by korydraughn**
This is a WIP.

It will be updated as 0.5.0 issues are resolved.

Leaving as draft for now. Feel free to review though.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+- Log IP address of clients (#73).
+- Remove experimental from package name (#159).
+- Distribute long-running write operations across thread-pool (#196).
+- Replace support for GenQuery2 API plugin with native iRODS 4.3.2 implementation (#272).
+- Improve documentation (#285, #286, #294, #309, #310, #312, #323).
+- Improve testing (#287).
+- Stat of resource no longer interprets resource status string (#299).
+- Remove unnecessary log functions (#324).
+- Rename namespace alias for logging to avoid naming collisions with C++ standard library (#325).
```
Comment (korydraughn) [inline CHANGELOG.md:49]: Consider rewording this line.

File: CHANGELOG.md [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+- Implement user-mapping plugin system (#293).
+
+## [0.4.0] - 2024-08-26
+
+This release includes several enhancements including more exposure of the iRODS API, logging of IP addresses, an improved multipart/form-data parser, and improved stability.
+
+### Changed
+
+- Log IP address of clients (#73).
+- Remove experimental from package name (#159).
```
Comment (korydraughn) [inline CHANGELOG.md:42]: Highlight word: experimental

File: CHANGELOG.md (hunk 1 above)
Comment (korydraughn) [reply CHANGELOG.md]: Done.

File: CHANGELOG.md (hunk 2 above)
Comment (korydraughn) [reply CHANGELOG.md]: Done.

File: CHANGELOG.md [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+## [0.5.0] - 2024-11-XX
+
+This release includes a user-mapping plugin system that enables dynamic user-mapping between OpenID Connect users and iRODS users. It also includes improved security through token validation.
+
+User-mapping functionality from previous versions of the HTTP API is now provided through the user-mapping plugin system. See [Mapping OpenID Users to iRODS](https://github.com/irods/irods_client_http_api/tree/0.5.0?tab=readme-ov-file#mapping-openid-users-to-irods) for details.
+
+### Changed
+
+- Validate OIDC ID tokens in client mode (#107).
```
Comment (trel) [inline CHANGELOG.md:22]: should this be in the 'added' section?

**Review by trel — COMMENTED** [review]
yes, looking ready to squash at least...

File: CHANGELOG.md (hunk 3 above)
Comment (korydraughn) [reply CHANGELOG.md:22]: Will move it.

Comment (korydraughn) [reply CHANGELOG.md:22]: After reading the changelog section descriptions again, I'm now thinking all validation commits should go under "Changed".

According to the changelog website:
- "Added" is for new features
- "Changed" is for changes in existing functionality

The validation commits change the behavior of the HTTP API, but for the user, there's no new functionality.

Thoughts?

Comment (trel) [reply CHANGELOG.md:22]: if validation is a feature, it was added.

if http api is doing auth differently... it is changed.

still feels like added to me... but your call.

Comment (korydraughn) [reply CHANGELOG.md:22]: That's true. Dang.

Going to have to use more brain power now.

Comment (korydraughn) [reply CHANGELOG.md:22]: @MartinFlores751 Where would you put the following issues given the descriptions at https://github.com/irods/irods_client_http_api/pull/351#discussion_r1840312028?
- 107
- 270
- 343

**Review by MartinFlores751 — COMMENTED** [review]
Words.

**Comment by trel** [comment]
looking good - please squash and resolve.

File: CHANGELOG.md [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+- Improve documentation for `tls_certificates_directory` OIDC property (#243).
+- Provide clear message on OIDC connection error (#244).
+- Collapse management of version number to single point of control (#269).
+- Improve handling of inaccessible configuration file on startup (#337).
+- Rename `realm_secret` to `access_token_secret` (#359).
```
[the remark this answers is not in the document]
Comment (korydraughn) [reply CHANGELOG.md:26]: Done.

**Comment by korydraughn** [comment]
Squashed and bumped version to 0.5.0.

Confirmed the package and the binary report include 0.5.0.

File: CHANGELOG.md [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+### Changed
+
+- Log IP address of clients (#73).
+- Remove **experimental** from package name (#159).
+- Distribute long-running write operations across thread-pool (#196).
+- Replace support for GenQuery2 API plugin with native iRODS 4.3.2 implementation (#272).
+- Improve documentation (#285, #286, #294, #309, #310, #312, #323).
+- Improve testing (#287).
+- Stat of resource no longer interprets resource status string (#299).
+- Remove unnecessary log functions (#324).
```
Comment (trel) [inline CHANGELOG.md:56]: is this is a remove?

File: CHANGELOG.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (alanking) [reply CHANGELOG.md:26]: I still see the line about renaming `realm_secret` to `access_token_secret`, despite `realm_secret` not existing in any released versions.

File: CHANGELOG.md [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+User-mapping functionality from previous versions of the HTTP API is now provided through the user-mapping plugin system. See [Mapping OpenID Users to iRODS](https://github.com/irods/irods_client_http_api/tree/0.5.0?tab=readme-ov-file#mapping-openid-users-to-irods) for details.
+
+### Changed
+
+- Improve documentation for `tls_certificates_directory` OIDC property (#243).
+- Provide clear message on OIDC connection error (#244).
+- Collapse management of version number to single point of control (#269).
+- Improve handling of inaccessible configuration file on startup (#337).
+- Rename `realm_secret` to `access_token_secret` (#359).
+- Leave space for null-terminating byte in destination buffer when using `std::string::copy` (#365).
```
Comment (alanking) [inline CHANGELOG.md:27]: Arguably a "Fixed"?

Comment (korydraughn) [reply CHANGELOG.md:27]: That's correct. Moving it.

File: CHANGELOG.md (hunk 5 above)
Comment (trel) [reply CHANGELOG.md:56]: okay

File: CHANGELOG.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (korydraughn) [reply CHANGELOG.md:26]: That's true. Will combine them.

**Comment by trel** [comment]
looking good, please squash again to two.

File: CHANGELOG.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (korydraughn) [reply CHANGELOG.md]: Done.

File: CHANGELOG.md (hunk 6 above)
Comment (korydraughn) [reply CHANGELOG.md]: Done.

**Comment by korydraughn** [comment]
Squashed.

**Comment by trel** [comment]
i think it's ready for #, awaiting alan vote.

**Pull request merged**